src/index/CompressedRelation.cpp

```diff
@@ -1599,15 +1659,31 @@ auto CompressedRelationWriter::createPermutationPair(
     for (size_t idx : ad_utility::integerRange(block.numRows())) {
       Id col0Id = firstCol[idx];
       decltype(auto) curRemainingCols = permutedCols[idx];
+
       if (col0Id != col0IdCurrentRelation) {
         finishRelation();
         col0IdCurrentRelation = col0Id;
       }
+
+      // Check if we need to create a new block before adding the current
+      // triple. We create a new block if:
+      // 1. The relation buffer is at the block size limit, AND
+      // 2. The current triple has different first three columns than the last
+      //    triple in the buffer (to ensure equal triples stay in same block)
+      if (relation.size() >= blocksize && !relation.empty()) {
+        // Compare first three columns of current triple with last buffered
+        // triple
+        const auto& lastBufferedRow = relation.back();
+        if (curRemainingCols[0] != lastBufferedRow[0] ||
+            curRemainingCols[1] != lastBufferedRow[1] ||
+            curRemainingCols[2] != lastBufferedRow[2]) {
```

Comment:
This pattern is super common in the code, I'm sure a helper function would already help a lot here.